### PR TITLE
 Speed-up callbacks, issue #4837 

### PR DESCRIPTION
**Pull request by shehr7**
* Resolves: #4837
* Target version: master 

### Summary
Changes:

    condition on _newsockets and _newcallbacks --> Inside wakeup function
    Use loop to read Wakeup pipe inside poll function --> in socket.cpp 

### TODO

- [ ] ...

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [x] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: net/Socket.cpp [hunk 1]
```diff
@@ -365,13 +365,20 @@ int SocketPoll::poll(int64_t timeoutMaxMicroS)
...
     {
         // Clear the data.
+        //Size of wakeup array
+        size_t arraySize = sizeof(_wakeup) / sizeof(_wakeup[0]);
 #if !MOBILEAPP
         int dump[32];
-        dump[0] = ::read(_wakeup[0], &dump, sizeof(dump));
-        LOG_TRC("Wakup pipe read " << dump[0] << " bytes");
+        //Empty wakeup pipe inside poll function.
+        for (size_t ii = 0; ii < arraySize; ii++)
```
Comment (GulsahKose) [inline net/Socket.cpp:373]: Here you can use "i" instead of "ii". Scope will be in loop so don't need to concern about it.

File: net/Socket.hpp [hunk 2]
```diff
@@ -715,8 +715,11 @@ class SocketPoll
             LOG_WRN("Waking up dead poll thread ["
                     << _name << "], started: " << (_threadStarted ? "true" : "false")
                     << ", finished: " << _threadFinished);
-
-        wakeup(_wakeup[1]);
+        // Check if _newSockets and _newCallbacks is empty before doing a 'wakeup()'
+        if (!_newSockets.empty() || !_newCallbacks.empty())
+        {
+            wakeup(_wakeup[1]);
```
Comment (GulsahKose) [inline net/Socket.hpp:721]: Here you can remove the curly braces too.

**Review by GulsahKose — COMMENTED** [review]
Hi Shehr,

Thanks for the fix and comments. Logic seems good to me. Just I can say something about naming etc. You can see my comments in lines.

File: net/Socket.cpp [hunk 3]
```diff
@@ -365,13 +365,20 @@ int SocketPoll::poll(int64_t timeoutMaxMicroS)
...
+        {
+            dump[0] = ::read(_wakeup[ii], &dump, sizeof(dump));
+            LOG_TRC("Wakup pipe read " << dump[0] << " bytes");
+        }
 #else
         LOG_TRC("Wakeup pipe read");
-        int dump = fakeSocketRead(_wakeup[0], &dump, sizeof(dump));
+        for (size_t ii = 0; ii < arraySize; ii++) {
+            int dump = fakeSocketRead(_wakeup[ii], &dump, sizeof(dump)); }
```
Comment (GulsahKose) [inline net/Socket.cpp:381]: Same. please use i instead of ii

**Comment by GulsahKose** [comment]
Unfortuanately i detected performance decline. You can test it in Writer. Just try to select one word. You'll see, there is some delay. I'll review more detailed and let you know.

**Comment by shehr7** [comment]
@GulsahKose I'm sorry, for I ran into some errors and had to update the file. I will let you know once the commit is up to date

**Comment by shehr7** [comment]
@GulsahKose I am not sure about this issue as I tested in calc and writer, and experienced a better performance. Might be because of some other reason.

**Comment by shehr7** [comment]
@Ashod @GulsahKose  I have made some amendments. Kindly check to see if it works, for me, this version works faster than the previous ones.

**Pull request closed without merging**